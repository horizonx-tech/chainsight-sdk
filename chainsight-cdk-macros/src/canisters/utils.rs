use std::path::PathBuf;

pub fn extract_contract_name_from_path(s: &str) -> String {
    let path = PathBuf::from(s);
    let name = path.file_stem().expect("file_stem failed");
    name.to_str().expect("to_str failed").to_string()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_extract_contract_name_from_path() {
        let path = "__interfaces/Oracle.json";
        assert_eq!(extract_contract_name_from_path(path), "Oracle");
    }
}
