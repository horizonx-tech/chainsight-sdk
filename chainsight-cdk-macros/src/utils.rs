use proc_macro::TokenStream;
use quote::quote;
use syn::parse_macro_input;

pub fn chainsight_common(input: TokenStream) -> TokenStream {
    let item = parse_macro_input!(input as syn::LitInt);
    let output = quote! {
        #[derive(Clone, Debug, PartialEq, candid::CandidType, candid::Deserialize, serde::Serialize)]
        pub struct CanisterMetricsSnapshot {
            pub timestamp: u64,
            pub cycles: u128,
        }
        chainsight_cdk_macros::manage_vec_state!("canister_metrics_snapshot", CanisterMetricsSnapshot, false);

        fn setup_monitoring_canister_metrics() {
            let round_timestamp = |ts: u32, unit: u32| ts / unit * unit;
            let current_time_sec = (ic_cdk::api::time() / (1000 * 1000000)) as u32;
            let delay = round_timestamp(current_time_sec, #item) + #item - current_time_sec;

            ic_cdk_timers::set_timer(std::time::Duration::from_secs(delay as u64), move || {
                ic_cdk_timers::set_timer_interval(std::time::Duration::from_secs(#item as u64), || {
                    monitor_canister_metrics();
                });
            });
            // in first time, immediate execution
            monitor_canister_metrics();
        }

        fn monitor_canister_metrics() {
            let timestamp = ic_cdk::api::time();
            let cycles = ic_cdk::api::canister_balance128();
            let datum = CanisterMetricsSnapshot {
                timestamp,
                cycles,
            };
            ic_cdk::println!("monitoring: {:?}", datum.clone());
            add_canister_metrics_snapshot(datum);
        }

        #[ic_cdk_macros::init]
        fn init() {
            setup_monitoring_canister_metrics()
        }

        #[ic_cdk_macros::post_upgrade]
        fn post_upgrade() {
            setup_monitoring_canister_metrics()
        }

        #[ic_cdk::query]
        #[candid::candid_method(query)]
        pub fn metric() -> CanisterMetricsSnapshot {
            get_last_canister_metrics_snapshot()
        }

        #[ic_cdk::query]
        #[candid::candid_method(query)]
        pub fn metrics(n: usize) -> Vec<CanisterMetricsSnapshot> {
            get_top_canister_metrics_snapshots(n)
        }

        async fn _get_target_proxy(target: candid::Principal) -> candid::Principal {
            let out: ic_cdk::api::call::CallResult<(candid::Principal,)> = ic_cdk::api::call::call(target, "get_proxy", ()).await;
            out.unwrap().0
        }
    };
    output.into()
}

pub fn did_export(input: TokenStream) -> TokenStream {
    let item = parse_macro_input!(input as syn::LitStr);
    let file_name = item.value() + ".did";
    TokenStream::from(quote! {
        candid::export_service!();
        #[ic_cdk::query(name = "__get_candid_interface_tmp_hack")]
        #[candid::candid_method(query, rename = "__get_candid_interface_tmp_hack")]
        fn __export_did_tmp_() -> String {
            __export_service()
        }
        #[cfg(test)]
        mod tests {
            use super::*;

            #[test]
            fn gen_candid() {
                std::fs::write(#file_name, __export_service()).unwrap();
            }
        }
    })
}
