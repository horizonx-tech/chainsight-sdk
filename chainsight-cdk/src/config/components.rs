use std::collections::HashMap;

use anyhow::bail;
use lazy_static::lazy_static;
use regex::Regex;

use crate::indexer::IndexingConfig;

lazy_static! {
    static ref MAPPING_CANDID_TY: HashMap<&'static str, &'static str> = [
        ("text", "String"),
        // ("blob", "&[u8]"),
        ("nat", "u128"),
        ("int", "i128"),
        ("nat8", "u8"),
        ("nat16", "u16"),
        ("nat32", "u32"),
        ("nat64", "u64"),
        ("int8", "i8"),
        ("int16", "i16"),
        ("int32", "i32"),
        ("int64", "i64"),
        ("float32", "f32"),
        ("float64", "f64"),
        ("bool", "bool"),
        // ("null", "()"),
    ].iter().cloned().collect();

    static ref REGEX_CANDID_FUNC: Regex = Regex::new(r"(?P<identifier>\w+)\s*:\s*\((?P<params>.*?)\)\s*(->\s*\((?P<return>.*?)\))?").unwrap();
    static ref REGEX_VECTOR: Regex = Regex::new(r"vec\s(?P<type>\w+)").unwrap();
    static ref REGEX_TUPLE: Regex = Regex::new(r"record\s\{\s(?P<items>(\w+(;\s|))+)\s\}").unwrap();
    static ref REGEX_STRUCT: Regex = Regex::new(r"(?P<field>\w+)\s*:\s*(?P<type>\w+)").unwrap();

    static ref REGEX_MULTIPLE_RECORD_TYPE: Regex = Regex::new(r"record\s*\{").unwrap();
}

#[derive(serde::Deserialize, serde::Serialize)]
pub struct CommonConfig {
    pub monitor_duration: u32,
    pub canister_name: String,
}

#[derive(serde::Deserialize, serde::Serialize)]
pub struct AlgorithmIndexerInput {
    pub method_name: String,
    pub response_type: String,
    pub source_type: AlgorithmInputType,
}

#[derive(Clone, Debug, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum AlgorithmInputType {
    #[serde(rename = "event_indexer")]
    EventIndexer,
    #[serde(rename = "key_value")]
    KeyValue,
    #[serde(rename = "key_values")]
    KeyValues,
}

impl Default for AlgorithmIndexerInput {
    fn default() -> Self {
        Self {
            method_name: "get_list".to_string(),
            response_type: "String".to_string(),
            source_type: AlgorithmInputType::EventIndexer,
        }
    }
}

impl Default for CommonConfig {
    fn default() -> Self {
        Self {
            monitor_duration: 3600,
            canister_name: "example_canister".to_string(),
        }
    }
}

#[derive(Default, serde::Serialize, serde::Deserialize)]
pub struct AlgorithmIndexerConfig {
    pub common: CommonConfig,
    pub indexing: IndexingConfig,
    pub input: AlgorithmIndexerInput,
}

#[derive(Default, serde::Serialize, serde::Deserialize)]
pub struct AlgorithmLensConfig {
    pub common: CommonConfig,
    pub target_count: usize,
}

#[derive(Default, serde::Serialize, serde::Deserialize)]
pub struct EventIndexerConfig {
    pub common: CommonConfig,
    pub def: EventIndexerEventDefinition,
}
#[derive(Default, Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct EventIndexerEventDefinition {
    pub identifier: String,
    pub abi_file_path: String,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct RelayerConfig {
    pub common: CommonConfig,
    pub destination: String,
    pub oracle_type: String,
    pub method_name: String,
    pub canister_method_value_type: CanisterMethodValueType,
    pub abi_file_path: String,
    pub lens_targets: Option<LensTargets>,
}
#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct LensTargets {
    pub identifiers: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
pub enum CanisterMethodValueType {
    Scalar(String, bool),                // struct name, is_scalar
    Tuple(Vec<(String, bool)>),          // struct_name, is_scalar
    Struct(Vec<(String, String, bool)>), // temp: Only non-nested `record` are supported.
    Vector(String, bool),                // struct_name, is_scalar
}
#[derive(Debug, Clone, PartialEq)]
pub struct CanisterMethodIdentifier {
    pub identifier: String,
    pub params: Vec<String>,
    pub return_value: CanisterMethodValueType,
}

impl CanisterMethodIdentifier {
    pub fn parse_from_str(s: &str) -> anyhow::Result<Self> {
        let captures = REGEX_CANDID_FUNC.captures(s).ok_or(anyhow::anyhow!(
            "method.identifier does not satisfy the supported expression: {}, supplied={}",
            REGEX_CANDID_FUNC.to_string(),
            s
        ))?;

        let identifier = captures.name("identifier").unwrap().as_str().to_string();

        let params_str = captures.name("params").unwrap().as_str();
        let params_result: Vec<String> = if params_str.is_empty() {
            vec![]
        } else {
            params_str
                .split(',')
                .map(|s| convert_type_from_candid_type(s.trim()).0)
                .collect()
        };
        let params = params_result;

        let return_value_str = captures.name("return").unwrap().as_str();
        let return_value = Self::parse_return_value(return_value_str)?;

        Ok(CanisterMethodIdentifier {
            identifier,
            params,
            return_value,
        })
    }

    fn parse_return_value(s: &str) -> anyhow::Result<CanisterMethodValueType> {
        let record_type_count = REGEX_MULTIPLE_RECORD_TYPE.find_iter(s).count();
        if record_type_count >= 2 {
            bail!(
                "Sorry, Currently nested `record` types are not supported. This will be supported in the future.\nTarget literal = {}",
                s
            ); // TODO: Support nested `record` types.
        }
        // vector
        if s.starts_with("vec") {
            let captures = REGEX_VECTOR.captures(s);
            if let Some(captures_value) = captures {
                let ty = captures_value.name("type").unwrap().as_str();
                let val = convert_type_from_candid_type(ty);
                return Ok(CanisterMethodValueType::Vector(val.0, val.1));
            }
            bail!("Invalid candid's result types:{}", s);
        }

        // Scalar
        if !s.starts_with("record") {
            let val = convert_type_from_candid_type(s);
            return Ok(CanisterMethodValueType::Scalar(val.0, val.1));
        }

        // Tuple
        let captures = REGEX_TUPLE.captures(s);
        if let Some(captures_value) = captures {
            let items = captures_value.name("items").unwrap().as_str();
            let tuple_result: Vec<(String, bool)> = items
                .split(';')
                .map(|s| convert_type_from_candid_type(s.trim()))
                .collect();
            let tuple = tuple_result;
            return Ok(CanisterMethodValueType::Tuple(tuple));
        }

        // Struct
        let items = REGEX_STRUCT.captures_iter(s);
        let mut struct_items = vec![];
        for cap in items {
            let field = cap.name("field").unwrap().as_str().to_string();
            let ty = convert_type_from_candid_type(cap.name("type").unwrap().as_str());
            struct_items.push((field, ty.0, ty.1));
        }
        if struct_items.is_empty() {
            bail!("Invalid candid's result types: {}", s);
        }
        Ok(CanisterMethodValueType::Struct(struct_items))
    }
}

fn convert_type_from_candid_type(s: &str) -> (String, bool) {
    // ref: https://internetcomputer.org/docs/current/references/candid-ref
    let ty_str = MAPPING_CANDID_TY.get(&s);
    if let Some(ty_str) = ty_str {
        return (ty_str.to_string(), true);
    }
    (s.to_string(), false)
}
