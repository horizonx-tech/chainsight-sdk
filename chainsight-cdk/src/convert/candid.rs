use std::{cmp::max, fs, ops::Deref, path::Path};

use anyhow::Ok;
use candid::{
    bindings::rust::{compile, Target},
    check_prog,
    types::{Type, TypeInner},
    IDLProg, TypeEnv,
};
use regex::Regex;

#[derive(Debug, Clone)]
pub struct CanisterMethodIdentifier {
    pub identifier: String,
    type_env: TypeEnv,
}
impl CanisterMethodIdentifier {
    pub const REQUEST_ARGS_TYPE_NAME: &'static str = "RequestArgsType";
    pub const RESPONSE_TYPE_NAME: &'static str = "ResponseType";
    pub const SUFFIX_IN_DID: &'static str = "InDid";

    pub fn new(s: &str) -> anyhow::Result<Self> {
        Self::new_internal(s, None)
    }

    pub fn new_with_did(s: &str, dependended_did: String) -> anyhow::Result<Self> {
        // NOTE: If the reserved Type specified in the .did to import is used, a Compile error will occur due to duplication.
        let (s, dependended_did) = avoid_using_reserved_types(s, &dependended_did);
        Self::new_internal(&s, Some(dependended_did))
    }

    fn new_internal(s: &str, dependended_did: Option<String>) -> anyhow::Result<Self> {
        let (identifier, args_ty, response_ty) = extract_elements(s)?;
        let did: String = Self::generate_did(&args_ty, &response_ty);

        let ast: IDLProg = if let Some(base_did) = dependended_did {
            format!("{}\n\n{}", base_did, did)
        } else {
            did.to_string()
        }
        .parse()?;
        let mut type_env = TypeEnv::new();
        let _ = check_prog(&mut type_env, &ast);

        Ok(Self {
            identifier,
            type_env,
        })
    }

    pub fn compile(&self) -> anyhow::Result<String> {
        anyhow::ensure!(self.compilable(), "Not compilable IDLProg");

        let mut config = candid::bindings::rust::Config::new();
        // // Update the structure derive to chainsight's own settings
        config.type_attributes = "#[derive(Clone, Debug, candid :: CandidType, candid :: Deserialize, serde :: Serialize, chainsight_cdk_macros :: StableMemoryStorable)]".to_string();
        config.target = Target::CanisterStub;
        let contents = compile(&config, &self.type_env, &None);

        let mut lines = contents
            .lines()
            // Delete initial 'use' declarations
            .skip(4)
            // Delete comment lines and blank lines
            .filter(|line| !(line.starts_with("//") || line.is_empty()))
            .map(|line| {
                // convert icp's own Nat and Int types to rust native type
                // NOTE: num-traits can be used, but is not used to reduce dependencies
                //  https://forum.dfinity.org/t/candid-nat-to-u128/16016
                //  https://discord.com/channels/748416164832608337/872791506853978142/1162494173933481984
                // NOTE: when ready to convert to u128/i128, consider with EthAbiEncoder's Encoder trait
                if line.contains("candid::Nat") {
                    return line.replace("candid::Nat", "u128");
                }
                if line.contains("candid::Int") {
                    return line.replace("candid::Int", "i128");
                }
                line.to_string()
            })
            .collect::<Vec<_>>();
        lines.insert(0, "#![allow(dead_code, unused_imports)]".to_string());
        lines.insert(
            1,
            "use candid::{self, CandidType, Deserialize, Principal, Encode, Decode};".to_string(),
        );
        Ok(lines.join("\n"))
    }

    pub fn get_types(&self) -> (Option<&Type>, Option<&Type>) {
        (
            self.find_type(Self::REQUEST_ARGS_TYPE_NAME),
            self.find_type(Self::RESPONSE_TYPE_NAME),
        )
    }

    pub fn get_type(&self, key: &str) -> Option<&Type> {
        self.find_type(key)
    }

    fn compilable(&self) -> bool {
        let (args_ty, response_ty) = self.get_types();
        let not_compilable_type = &TypeInner::Unknown;

        if args_ty.is_some_and(|ty| ty.deref().eq(not_compilable_type)) {
            return false;
        }
        if response_ty.is_some_and(|ty| ty.deref().eq(not_compilable_type)) {
            return false;
        }

        true
    }

    fn generate_did(args_ty: &str, response_ty: &str) -> String {
        let args_ty_did = if args_ty.is_empty() {
            "".to_string()
        } else {
            generate_did_type(Self::REQUEST_ARGS_TYPE_NAME, args_ty)
        };
        let response_ty_did = if response_ty.is_empty() {
            "".to_string()
        } else {
            generate_did_type(Self::RESPONSE_TYPE_NAME, response_ty)
        };
        format!("{}\n{}", args_ty_did, response_ty_did)
    }

    fn find_type(&self, key: &str) -> Option<&Type> {
        let ty = self.type_env.find_type(key);
        ty.ok()
    }
}

fn generate_did_type(key: &str, value: &str) -> String {
    format!("type {} = {};", key, value)
}

fn extract_elements(s: &str) -> anyhow::Result<(String, String, String)> {
    let (identifier, remains) = s
        .split_once(':')
        .expect("Invalid canister method identifier");
    let (args_ty, response_ty) = remains
        .split_once("->")
        .expect("Invalid canister method identifier");

    let trim_type_str = |s: &str| {
        let trimed = s.trim();
        let removed_brackets = trimed.trim_matches(|c| c == '(' || c == ')');
        removed_brackets.trim().to_string()
    };

    Ok((
        identifier.trim().to_string(),
        trim_type_str(args_ty),
        trim_type_str(response_ty),
    ))
}

fn avoid_using_reserved_types(s: &str, did: &str) -> (String, String) {
    let req_ty = CanisterMethodIdentifier::REQUEST_ARGS_TYPE_NAME;
    let res_ty = CanisterMethodIdentifier::RESPONSE_TYPE_NAME;
    let base_suffix = CanisterMethodIdentifier::SUFFIX_IN_DID;

    // Check maximum value of suffix for reserved type
    let pattern = format!(
        r"(?P<name>({}|{}))_{}_(?P<suffix_num>\d+)",
        req_ty, res_ty, base_suffix
    );
    let re = Regex::new(&pattern).expect("Invalid regex pattern");
    let mut max_number = 0;
    for line in did.lines() {
        for cap in re.captures_iter(line) {
            let suffix_num = cap.name("suffix_num").unwrap().as_str();
            if let core::result::Result::Ok(number) = suffix_num.parse::<u32>() {
                max_number = max(max_number, number);
            }
        }
    }
    max_number += 1;

    // Replace reserved type (did)
    let pattern = format!(r" (?P<name>({}|{}))(?P<last_char>(;|,| ))", req_ty, res_ty);
    let re = Regex::new(&pattern).expect("Invalid regex pattern");
    let replace_reserved_type = |s: &str| {
        let replaced_s = re.replace_all(s, |caps: &regex::Captures| {
            let name = caps.name("name").unwrap().as_str();
            let last_char = caps.name("last_char").unwrap().as_str();
            format!(" {}_{}_{}{}", name, base_suffix, max_number, last_char)
        });
        replaced_s.to_string()
    };
    let replaced_did_lines = did.lines().map(replace_reserved_type).collect::<Vec<_>>();

    // Replace reserved type (identifier)
    let pattern = format!(
        r"(?P<first_char>(\(| ))(?P<name>({}|{}))(?P<last_char>(\)|;|,| ))",
        req_ty, res_ty
    );
    let re = Regex::new(&pattern).expect("Invalid regex pattern");
    let replaced_s = re.replace_all(s, |caps: &regex::Captures| {
        let first_char = caps.name("first_char").unwrap().as_str();
        let name = caps.name("name").unwrap().as_str();
        let last_char = caps.name("last_char").unwrap().as_str();
        format!(
            "{}{}_{}_{}{}",
            first_char, name, base_suffix, max_number, last_char
        )
    });

    (replaced_s.to_string(), replaced_did_lines.join("\n"))
}

// Read .did and remove 'service' section
pub fn read_did_to_string_without_service<P: AsRef<Path>>(path: P) -> anyhow::Result<String> {
    let mut data = fs::read_to_string(path)?;
    Ok(exclude_service_from_did_string(&mut data))
}
fn exclude_service_from_did_string(data: &mut String) -> String {
    // remove 'service' section
    if let Some(service_start_idx) = data.find("service : {") {
        data.truncate(service_start_idx);
    }

    data.trim().to_owned()
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_snapshot;

    const TEST_IDENTS: &'static [(&'static str, &'static str); 6] = &[
        (&"normal", &"update_value : (nat64) -> (text)"),
        (&"normal_nat", &"use_nat : (nat32) -> (nat)"),
        (&"normal_int", &"use_nat : (int32) -> (int)"),
        (&"no_args", &"get_value : () -> (text)"),
        (
            &"record",
            &"get_snapshot : (nat64) -> (record { value : text; timestamp : nat64 })",
        ),
        (
            &"nested_record",
            &"get_last_snapshot : () -> (record { dai : record { usd : float64 } })",
        ),
    ];

    const TEST_IDENTS_WITH_DID: &'static [(&'static str, &'static str, &'static str); 5] = &[
        (
            &"single",
            &"get_snapshot : (nat64) -> (Snapshot)",
            &"type Snapshot = record { value : text; timestamp : nat64 };",
        ),
        (
            &"multiple",
            &"get_snapshot : (nat64) -> (Snapshot)",
            &r#"type CurrencyValue = record { usd : float64 };
type Snapshot = record { value : SnapshotValue; timestamp : nat64 };
type SnapshotValue = record { dai : CurrencyValue };
"#,
        ),
        (
            &"with_variant",
            &"init_in : (Env) -> (Result)",
            &r#"type Env = variant { Production; Test; LocalDevelopment };
type InitError = variant {
  InvalidDestination : text;
  InvalidPrincipal : principal;
  InvalidContent : text;
  InvalidRequest : text;
};
type Result = variant { Ok; Err : InitError };
"#,
        ),
        (
            &"with_vec",
            &"get_sources : () -> (vec Sources)",
            &r#"type HttpsSnapshotIndexerSourceAttrs = record {
  queries : vec record { text; text };
};
type SourceType = variant { evm; https; chainsight };
type Sources = record {
  source : text;
  interval_sec : opt nat32;
  attributes : HttpsSnapshotIndexerSourceAttrs;
  source_type : SourceType;
};"#,
        ),
        (
            &"with_reserved_type",
            &"get_snapshot : (RequestArgsType) -> (ResponseType)",
            &r#"type RequestArgsType = nat64;
type ResponseType = text;"#,
        ),
    ];

    #[test]
    fn test_compile() {
        for (label, s) in TEST_IDENTS {
            let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
            let compiled = ident.compile().unwrap();
            assert_snapshot!(format!("compile__{}", label.to_string()), compiled);
        }
    }

    #[test]
    fn test_compile_with_depended_did() {
        for (label, s, did) in TEST_IDENTS_WITH_DID {
            let ident = CanisterMethodIdentifier::new_with_did(s, did.to_string())
                .expect("Failed to parse");
            let compiled = ident.compile().unwrap();
            assert_snapshot!(
                format!("compile_with_depended_did__{}", label.to_string()),
                compiled
            );
        }
    }

    #[test]
    fn test_compile_when_not_compilable() {
        let s = "get_snapshot : (Input) -> (Output)";
        let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
        assert_eq!(
            ident.compile().err().unwrap().to_string(),
            "Not compilable IDLProg"
        );
    }

    #[test]
    fn test_compilable() {
        // compilable
        for (_, s) in TEST_IDENTS {
            let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
            assert!(ident.compilable());
        }
        for (_, s, did) in TEST_IDENTS_WITH_DID {
            let ident = CanisterMethodIdentifier::new_with_did(s, did.to_string())
                .expect("Failed to parse");
            assert!(ident.compilable());
        }

        // not compilable
        let not_compilable = vec![
            "get_snapshot : (Input) -> (Output)",
            "get_snapshot : (nat64) -> (Snapshot)",
            "get_snapshot : (Snapshot) -> (nat64)",
        ];
        for s in not_compilable {
            let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
            assert!(!ident.compilable());
        }

        let not_compilable_with_did = vec![
            (
                "get_snapshot : (Snapshot) -> (Snapshot)",
                "type Snapshot_1 = record { value : text; timestamp : nat64 };",
            ),
            (
                "get_snapshot : (nat64) -> (Snapshot)",
                "type Snapshot_1 = record { value : text; timestamp : nat64 };",
            ),
            (
                "get_snapshot : (Snapshot) -> (nat64)",
                "type Snapshot_1 = record { value : text; timestamp : nat64 };",
            ),
        ];
        for (s, did) in not_compilable_with_did {
            let ident = CanisterMethodIdentifier::new_with_did(s, did.to_string())
                .expect("Failed to parse");
            assert!(!ident.compilable());
        }
    }

    #[test]
    fn test_generate_did() {
        let key = "Account";
        let value = "record { owner : principal; subaccount : opt blob }";
        let did = generate_did_type(key, value);
        assert_snapshot!(did);
    }

    #[test]
    fn test_extract_elements() {
        for (label, s) in TEST_IDENTS {
            let result = extract_elements(s).expect("Failed to parse");
            assert_snapshot!(
                format!("extract_elements__{}", label.to_string()),
                format!("{:#?}", result)
            );
        }
    }

    #[test]
    fn test_extract_elements_multi_nested_record() {
        let s =
            "icrc2_allowance : (record { account : record { owner : principal; subaccount : opt blob }; spender : record { owner : principal; subaccount : opt blob } }) -> (record { allowance : nat; expires_at : opt nat64 })";
        let (identifier, args_ty, response_ty) = extract_elements(s).expect("Failed to parse");
        assert_eq!(identifier, "icrc2_allowance");
        assert_eq!(args_ty, "record { account : record { owner : principal; subaccount : opt blob }; spender : record { owner : principal; subaccount : opt blob } }");
        assert_eq!(
            response_ty,
            "record { allowance : nat; expires_at : opt nat64 }"
        );
    }

    #[test]
    fn test_exclude_service_from_did_string() {
        let expected = r#"type List = opt record { head: int; tail: List };
type byte = nat8;".to_string()"#;
        let mut actual = format!("{}\n", expected)
            + r#"service : {
  f : (byte, int, nat, int8) -> (List);
  g : (List) -> (int) query;
}"#;

        assert_eq!(exclude_service_from_did_string(&mut actual), expected);
    }
}
