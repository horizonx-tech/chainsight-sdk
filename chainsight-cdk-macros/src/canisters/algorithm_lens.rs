use chainsight_cdk::config::components::AlgorithmLensConfig;
use proc_macro::TokenStream;
use quote::{format_ident, quote};
use syn::parse_macro_input;

use crate::canisters::utils::update_funcs_to_upgrade;

pub fn def_algorithm_lens_canister(input: TokenStream) -> TokenStream {
    let input_json_string: String = parse_macro_input!(input as syn::LitStr).value();
    let config: AlgorithmLensConfig =
        serde_json::from_str(&input_json_string).expect("Failed to parse input_json_string");
    algorithm_lens_canister(config).into()
}

fn algorithm_lens_canister(config: AlgorithmLensConfig) -> proc_macro2::TokenStream {
    let canister_name = config.common.canister_name.clone();
    let canister_name_ident = format_ident!("{}", config.common.canister_name);
    let lens_size = config.target_count;
    let lens_method_quote = if let Some(args_type) = config.args_type {
        let args_type_ident = format_ident!("{}", args_type);
        quote! { lens_method!(#lens_size, #args_type_ident); }
    } else {
        quote! { lens_method!(#lens_size); }
    };

    let quote_to_upgradable = {
        let state_struct = quote! {
            #[derive(Clone, Debug, PartialEq, candid::CandidType, serde::Serialize, serde::Deserialize, CborSerde)]
            pub struct UpgradeStableState {
                pub initializing_state: InitializingState,
            }
        };

        let update_funcs_to_upgrade = update_funcs_to_upgrade(
            quote! {
                UpgradeStableState {
                    initializing_state: get_initializing_state()
                }
            },
            quote! {
                set_initializing_state(state.initializing_state);
            },
        );

        quote! {
            #state_struct
            #update_funcs_to_upgrade
        }
    };

    quote! {
        did_export!(#canister_name);
        use candid::{Decode, Encode};
        use chainsight_cdk_macros::{chainsight_common, did_export, init_in, lens_method, manage_single_state, stable_memory_for_scalar, prepare_stable_structure, StableMemoryStorable, CborSerde};
        use ic_stable_structures::writer::Writer;
        use ic_web3_rs::futures::{future::BoxFuture, FutureExt};
        chainsight_common!();
        init_in!();
        prepare_stable_structure!();
        use #canister_name_ident::*;
        #lens_method_quote
        #quote_to_upgradable
    }
}

#[cfg(test)]
mod test {
    use chainsight_cdk::config::components::CommonConfig;
    use insta::assert_snapshot;
    use rust_format::{Formatter, RustFmt};

    use super::*;

    #[test]
    fn test_snapshot_without_args() {
        let config = AlgorithmLensConfig {
            common: CommonConfig {
                canister_name: "app".to_string(),
            },
            target_count: 10,
            args_type: None,
        };
        let generated = algorithm_lens_canister(config);
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__algorithm_lens", formatted);
    }

    #[test]
    fn test_snapshot_with_args() {
        let config = AlgorithmLensConfig {
            common: CommonConfig {
                canister_name: "app".to_string(),
            },
            target_count: 10,
            args_type: Some("CalculateArgs".to_string()),
        };
        let generated = algorithm_lens_canister(config);
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__algorithm_lens_with_args", formatted);
    }
}
