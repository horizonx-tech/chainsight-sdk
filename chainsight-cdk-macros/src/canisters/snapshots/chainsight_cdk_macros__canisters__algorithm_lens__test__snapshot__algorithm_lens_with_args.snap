---
source: chainsight-cdk-macros/src/canisters/algorithm_lens.rs
expression: formatted
---
did_export!("app");
use chainsight_cdk_macros::{chainsight_common, did_export, init_in, lens_method};
use ic_web3_rs::futures::{future::BoxFuture, FutureExt};
chainsight_common!(1000u32);
init_in!();
use app::*;
lens_method!(10usize, "CalculateArgs");

